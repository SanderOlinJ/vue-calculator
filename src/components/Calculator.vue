<template>
    <div id="grid-container">
        <div class="calculator-row" id="outer-calculator-div">
            <div id="calculator-output-div">
                <div id="calculator-output">
                    {{ updatingValue }}
                </div>
            </div>
        </div>
        <div class="calculator-row">
            <button class="calculator-button" v-for="n in calculatorFirstRow" :key="n"
                :class="{'pink-buttons': ['PWR','DEL','AC','รท'].includes(n)}" 
                @click="buttonClick(n)">
                {{ n }}
            </button>
        </div>
        
        <div class="calculator-row">
            <button class="calculator-button" v-for="n in calculatorSecondRow" :key="n"
                :class="{'green-buttons': ['7','8','9'].includes(n), 'pink-buttons': ['x'].includes(n)}"
                @click="buttonClick(n)">
                {{ n }}
            </button>
        </div>

        <div class="calculator-row">
            <button class="calculator-button" v-for="n in calculatorThirdRow" :key="n"
                :class="{'green-buttons': ['4','5','6'].includes(n), 'pink-buttons': ['-'].includes(n)}"
                @click="buttonClick(n)">
                {{ n }}
            </button>
        </div>

        <div class="calculator-row">
            <button class="calculator-button" v-for="n in calculatorFourthRow" :key="n"
                :class="{'green-buttons': ['1','2','3'].includes(n), 'pink-buttons': ['+'].includes(n)}"
                @click="buttonClick(n)">
                {{ n }}
            </button>
        </div>

        <div class="calculator-row">
            <div id="zero-button-div">
                <button class="calculator-button, green-buttons" id="zero-button"
                @click="buttonClick(0)">
                    0
                </button>
            </div>
            <button class="calculator-button" v-for="n in calculatorFifthRow" :key="n"
                :class="{'pink-buttons': ['.','='].includes(n)}"
                @click="buttonClick(n)">
                {{ n }}
            </button>
        </div>
    </div>
</template>

<script>
    export default{
        name: "Calculator",

        data() {
            return {
                updatingValue: "OFF",
                value: 0,
                lastOperator: "",
                equation: "",
                isOff: true,
                calculatorFirstRow: ["PWR","DEL", "AC", "รท"],
                calculatorSecondRow: ["7", "8", "9", "x"],
                calculatorThirdRow: ["4", "5", "6", "-"],
                calculatorFourthRow: ["1", "2", "3", "+"],
                calculatorFifthRow: [".", "="]
            }
        },

        methods: {
            buttonClick(buttonValue){
                if (buttonValue == "PWR") {
                    if (this.isOff){
                        this.clearOutAllFields();
                        this.isOff = false;
                        return;
                    } else{
                        this.clearOutAllFields();
                        this.updatingValue = "OFF";
                        this.isOff = true;
                        return;
                    }
                }
                if (this.isOff){
                    return;
                }
                if(isNaN(buttonValue) && buttonValue != "."){
                    this.symbolHandling(buttonValue);
                } else {
                    this.numberHandling(buttonValue);
                }
            },


            numberHandling(buttonValue){
                if(isNaN(this.updatingValue) && this.updatingValue != "."){
                    this.updatingValue = "";
                }
                this.updatingValue = this.updatingValue.concat(buttonValue);
            },

            symbolHandling(buttonValue){
                switch (buttonValue){
                    case "DEL":
                        this.clearOutAllFields();
                        break;
                    case "AC":
                        if (!this.updatingValue == ""){
                            this.updatingValue = this.updatingValue.slice(0, -1);
                        } else {
                            this.clearOutAllFields();
                        }
                        break;
                    case "=":
                        this.doTheMath(buttonValue);
                        this.updatingValue = this.value;
                        this.equation = this.equation.concat(this.value);
                        //this.$emit("addEquationToHistory", this.equation);
                        console.log(this.equation);
                        this.clearAllButOutput();
                        break;
                    default:
                        this.doTheMath(buttonValue);
                        break;
                }
            },

            doTheMath(buttonValue){
                console.log("--------")
                if(this.lastOperator != ""){
                    switch(this.lastOperator){
                        case "+":
                            this.value += parseInt(this.updatingValue);
                            break;
                        case "-":
                            this.value -= parseInt(this.updatingValue);
                            break;
                        case "x":
                            this.value *= parseInt(this.updatingValue);
                            break;
                        case "รท":
                            this.value /= parseInt(this.updatingValue);
                            break;
                    }
                } else {
                    if (this.updatingValue == ""){
                        this.value = 0;
                    } else {
                    this.value = parseInt(this.updatingValue);
                    }
                }
                if (this.updatingValue == ""){
                    this.equation = this.equation.concat("0");
                } else{
                    this.equation = this.equation.concat(this.updatingValue);
                }
                this.equation = this.equation.concat(buttonValue);
                this.updatingValue = buttonValue;
                this.lastOperator = buttonValue;
                
                console.log(this.equation);
                console.log(this.value);
                console.log(this.updatingValue);
                
            },

            clearOutAllFields(){
                this.clearAllButOutput();
                this.updatingValue = "";
            },

            clearAllButOutput(){
                this.value = 0;
                this.lastOperator = "";
                this.equation = "";
            }
        }
    }
</script>


<style>
@import url('https://fonts.googleapis.com/css2?family=Orbitron:wght@600&display=swap');
#grid-container{
    display: grid;
    color: white;
    height: 600px;
    width: 500px;
    grid-template-columns: repeat(auto, 1fr);
    grid-template-rows:  repeat(auto, 1fr);
    background-color: rgb(35, 35, 72);
    border: 10px solid black;
    border-radius: 10px;
}

.calculator-row{
    width: 500px;
    min-height: 88px;
    text-align: left;
    display: flex;
}

.calculator-button{
    min-width: 125px;
    height: 88px;
    font-size: 40px;
    border-radius: 5px;
    cursor: pointer;
}

.calculator-button:active{
    border: 3px solid black;
    font-size: 36px;
}

#calculator-output{
    padding-top: 12px;
    padding-right: 10px;
    height: 100px;
    text-align: right;
    justify-items: center;
    font-size: 70px;
    font-family: 'Orbitron', sans-serif;
    overflow-x: auto;
    overflow-y: hidden;
}

#calculator-output-div{
    width: 460px;
    height: 100px;
    border-radius: 20px;
    margin-left: 20px;
    margin-top: 25px;
    background-color: rgb(6, 6, 50);
}

#outer-calculator-div{
    height: 160px;
}

.pink-buttons{
    background-color: #fe67c2;
    border: 1px solid black;
}

.green-buttons{
    background-color: #62f2b4;
    border: 1px solid black;
}

#zero-button-div{
    display: grid;
    width: 250px;
    min-height: 88px;
}

#zero-button{
    font-size: 40px;
    border-radius: 5px;
    cursor: pointer;
}

#zero-button:active{
    border: 3px solid black;
    font-size: 36px;
}

</style>