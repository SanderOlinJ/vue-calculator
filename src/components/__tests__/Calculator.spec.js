import { describe, it, expect } from "vitest"
import { mount } from "@vue/test-utils"
import Calculator from '../Calculator.vue'

describe("Calculator", () => {
    const wrapper = mount(Calculator)
    it("Should render properly", () => {
        expect(wrapper.html()).toMatchSnapshot()
    })
    it("Should contain zero button", () => {
        expect(wrapper.find("zero-button")).toBeTruthy()
    })
})
