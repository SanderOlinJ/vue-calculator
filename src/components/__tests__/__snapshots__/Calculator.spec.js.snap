// Vitest Snapshot v1

exports[`Calculator > Should render properly 1`] = `
"<div id=\\"grid-container\\" data-v-453dfca7=\\"\\">
  <div class=\\"calculator-row\\" id=\\"outer-calculator-div\\" data-v-453dfca7=\\"\\">
    <div id=\\"calculator-output-div\\" data-v-453dfca7=\\"\\">
      <div id=\\"calculator-output\\" data-v-453dfca7=\\"\\">OFF</div>
    </div>
  </div>
  <div class=\\"calculator-row\\" data-v-453dfca7=\\"\\"><button class=\\"calculator-button pink-buttons\\" data-v-453dfca7=\\"\\">PWR</button><button class=\\"calculator-button pink-buttons\\" data-v-453dfca7=\\"\\">DEL</button><button class=\\"calculator-button pink-buttons\\" data-v-453dfca7=\\"\\">AC</button><button class=\\"calculator-button pink-buttons\\" data-v-453dfca7=\\"\\">รท</button></div>
  <div class=\\"calculator-row\\" data-v-453dfca7=\\"\\"><button class=\\"calculator-button green-buttons\\" data-v-453dfca7=\\"\\">7</button><button class=\\"calculator-button green-buttons\\" data-v-453dfca7=\\"\\">8</button><button class=\\"calculator-button green-buttons\\" data-v-453dfca7=\\"\\">9</button><button class=\\"calculator-button pink-buttons\\" data-v-453dfca7=\\"\\">x</button></div>
  <div class=\\"calculator-row\\" data-v-453dfca7=\\"\\"><button class=\\"calculator-button green-buttons\\" data-v-453dfca7=\\"\\">4</button><button class=\\"calculator-button green-buttons\\" data-v-453dfca7=\\"\\">5</button><button class=\\"calculator-button green-buttons\\" data-v-453dfca7=\\"\\">6</button><button class=\\"calculator-button pink-buttons\\" data-v-453dfca7=\\"\\">-</button></div>
  <div class=\\"calculator-row\\" data-v-453dfca7=\\"\\"><button class=\\"calculator-button green-buttons\\" data-v-453dfca7=\\"\\">1</button><button class=\\"calculator-button green-buttons\\" data-v-453dfca7=\\"\\">2</button><button class=\\"calculator-button green-buttons\\" data-v-453dfca7=\\"\\">3</button><button class=\\"calculator-button pink-buttons\\" data-v-453dfca7=\\"\\">+</button></div>
  <div class=\\"calculator-row\\" data-v-453dfca7=\\"\\">
    <div id=\\"zero-button-div\\" data-v-453dfca7=\\"\\"><button class=\\"calculator-button, green-buttons\\" id=\\"zero-button\\" data-v-453dfca7=\\"\\"> 0 </button></div><button class=\\"calculator-button pink-buttons\\" data-v-453dfca7=\\"\\">.</button><button class=\\"calculator-button pink-buttons\\" data-v-453dfca7=\\"\\">=</button>
  </div>
</div>"
`;
